mantidproject/mantid
# Implement EventList generateHistogram without sorting, use within AlignAndFocusPowder

**Description of work**

A new `generateHistogram` method has been added the `EventList` that will create the histogram without sorting the events first. Sorting events is slow, so this approach is faster. If the events are already sorted then the old method is called because it will be faster.

This should have a performance improvement when doing `Rebin` with `PreserveEvents=False` on any `EventWorkspace` that is unsorted. It does this by performing the histogram in a different way, it calculates which bin the event should go in instead of sorting the events first.

The `SortEvents` call in `AlignAndFocusPowder` for Ragged binning was removed to benefit from this speed improvement.

**To test:**

You can test this by running Rebin with `PreserveEvents=False` with any unsorted EventWorkspace.

To see an example of the performance improvements on my laptop, using `PG3_2538_event.nxs` from the SystemTest data and with different binning and number of histograms (by grouping them). The print at the end also shows that the input doesn't get sorted unlike what currently happens.

```python
for groupBy in (None, "bank", "All"):
    for step in ("100", "-0.01"):
        ws=LoadEventNexus("PG3_2538_event.nxs")
        if groupBy:
            groups = CreateGroupingWorkspace(ws, GroupDetectorsBy=groupBy)
            ws = GroupDetectors(ws, CopyGroupingFromWorkspace="groups")
        binned = Rebin(ws, f"9000,{step},25000", PreserveEvents=False)
        print(ws.getSpectrum(0).isSortedByTof())
```

|branch|linear|log|linear, group(bank)|log, group(bank)|linear, group(All)|log, group(All)|
|-|-|-|-|-|-|-|
|main|1.14s|1.03s|1.27s|1.27s|2.22s|2.24s|
|this|0.13s|0.34s|0.14s|0.37s|0.49s|1.24s|





Fixes [EWM1866](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=1866)








---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
